core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala

```diff
@@ -177,26 +174,36 @@ class TransactionCoordinator(brokerId: Int,
     }
   }
 
-  private def initProducerIdWithExistingMetadata(transactionalId: String,
-                                                 transactionTimeoutMs: Int,
-                                                 coordinatorEpoch: Int,
-                                                 txnMetadata: TransactionMetadata): Either[InitProducerIdResult, (Int, TxnTransitMetadata)] = {
+  private def prepareInitProduceIdTransit(transactionalId: String,
+                                          transactionTimeoutMs: Int,
+                                          coordinatorEpoch: Int,
+                                          txnMetadata: TransactionMetadata): Either[InitProducerIdResult, (Int, TxnTransitMetadata)] = {
     if (txnMetadata.pendingTransitionInProgress) {
       // return a retriable exception to let the client backoff and retry
       Left(initTransactionError(Errors.CONCURRENT_TRANSACTIONS))
     } else {
       // caller should have synchronized on txnMetadata already
       txnMetadata.state match {
         case PrepareAbort | PrepareCommit =>
-          // reply to client and let client backoff and retry
+          // reply to client and let it backoff and retry
           Left(initTransactionError(Errors.CONCURRENT_TRANSACTIONS))
 
         case CompleteAbort | CompleteCommit | Empty =>
-          // try to append and then update
-          Right(coordinatorEpoch, txnMetadata.prepareIncrementProducerEpoch(transactionTimeoutMs, time.milliseconds()))
+          val transitMetadata = if (txnMetadata.isProducerEpochExhausted) {
+            val newProducerId = producerIdManager.generateProducerId()
+            txnMetadata.prepareProducerIdRotation(newProducerId, transactionTimeoutMs, time.milliseconds())
+          } else {
+            txnMetadata.prepareIncrementProducerEpoch(transactionTimeoutMs, time.milliseconds())
+          }
+
+          Right(coordinatorEpoch, transitMetadata)
 
         case Ongoing =>
-          // indicate to abort the current ongoing txn first
+          // indicate to abort the current ongoing txn first. Note that this epoch is never returned to the
+          // user. We will abort the ongoing transaction and return CONCURRENT_TRANSACTIONS to the client.
+          // This forces the client to retry, which will ensure that the epoch is bumped a second time. In
+          // particular, if fencing the current producer exhausts the available epochs for the current producerId,
+          // then when the client retries, we will generate a new producerId.
```

Comment:
Yes, correct. Admittedly, it felt a little uneasy to depend on the fact that aborting an ongoing transaction in InitProducerId requires two epoch bumps, but perhaps we can improve on it in the next release. At least for now I try to add some defensive checks to make sure we don't accidentally "fix" this.